sympy/polys/ring_series.py

```diff
@@ -572,6 +574,110 @@ def _coefficient_t(p, t):
             p1[monomial_div(expv, expv1)] = p[expv]
     return p1
 
+def rs_series_reversion_newton(p, x, n):
+    r"""
+    Fast series reversion (aka Lagrange Inversion) using Newton-like updates.
+
+    Given a ring polynomial $p$ in variable $x$, this method computes a ring polynomial
+    $q$ in $x$ such that $q(p(x)) = x + O(x^n)$.
+
+    The series should be of form $p(x) = a_1 x + a_2 x^2 + ...$ where $a_1$ is invertible.
+
+    Parameters
+    ==========
+
+    p : :class:`~.PolyElement`
+        The polynomial to be inverted.
+    x : :class:`~.PolyElement`
+        The variable with respect to which ``p`` is inverted.
+    n : integer
+        The number of expansion terms to be computed.
+
+    Explanation
+    ===========
+
+    The algorithm adapts the Newton method to solve a general equation
+    $$
+    f(x)=y.
+    $$
+    By Taylor's expansion:
+    $$
+    f(x+h) = f(x) + f'(x) h + O(h^2).
+    $$
+    If $f(x) = y + O(y^{k+1})$ then with $h = -\frac{f(x)-y}{f'(x)} = O(y^{k+1})$ we obtain:
+    $$
+    f(x+h)-y = O(h^2) = O(y^{2k+2}).
+    $$
+    Thus, updating $x\gets x-h$ increases the accuracy from $k$ to $2k+1$ terms.
+
+    Examples
+    ========
+
+    In the example below we invert $f=\mathrm{e}^{x}-1$ around $x=0$:
+
+    >>> from sympy.polys.domains import QQ
+    >>> from sympy.polys.rings import ring
+    >>> from sympy.polys.ring_series import rs_exp, rs_series_reversion_newton
+    >>> Rx, xr = ring('x', QQ)
+    >>> n = 4
+    >>> fr = rs_exp(xr,xr,n)-1
+    >>> rs_series_reversion_newton(fr,xr,n)
+    1/3*x**3 - 1/2*x**2 + x
+
+    Another example derives general formulas for the inversion series:
+
+    >>> from sympy import IndexedBase, symbols
+    >>> from sympy.polys.ring_series import rs_series_reversion_newton
+    >>> a = IndexedBase('a')
+    >>> x = symbols('x')
+    >>> f = sum(a[k]*x**k for k in range(1,4))
+    >>> R = f.as_poly(x,field=True).domain
+    >>> Rx = R[x]
+    >>> fr = Rx.from_sympy(f)
+    >>> xr = Rx.from_sympy(x)
+    >>> rs_series_reversion_newton(fr,xr,4).as_expr()
+    x**3*(-a[3]/a[1]**4 + 2*a[2]**2/a[1]**5) - x**2*a[2]/a[1]**3 + x/a[1]
+
+
+    References
+    ==========
+
+    .. [1] Brent, Richard P., and Hsiang T. Kung. "Fast algorithms for manipulating formal power series." Journal of the ACM (JACM) 25, no. 4 (1978): 581-595.
+    .. [2] Weisstein, Eric W. "Series Reversion." From MathWorld--A Wolfram Web Resource. https://mathworld.wolfram.com/SeriesReversion.html
```

Comment:
Long lines should be wrapped.

Reply:
Changed in 05761dff628e4a030b40db5ea12f443376f5aeb8.